## BRAND NAME : FEMININO

<img src="C:/Users/ROSZ/Downloads/WhatsApp Image 2025-09-20 at 06.17.40.jpeg" width="1000">

## Importing Libraries

In [24]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# OBJECTIVES

* Analyzing the most effective marketing channels driving engagement and conversions, enabling the business to optimise campaigns,improve delivery performance and enhance customer experience.
* Evaluating target audience  most responsive to each marketing channel.
* Understanding ordering patterns across customer segments and areas.

## Loading Tables

In [25]:
feedback = pd.read_csv('blinkit_customer_feedback.csv')
customers = pd.read_csv('blinkit_customers.csv')
marketing = pd.read_csv('blinkit_marketing_performance.csv')
order_items = pd.read_csv('blinkit_order_items.csv')
orders = pd.read_csv('blinkit_orders.csv')
products = pd.read_csv('blinkit_products.csv')

In [26]:
##creating copy of the original tables
feedback_copy =feedback.copy() 
customers_copy =customers.copy() 
marketing_copy =marketing.copy()
order_items_copy =order_items.copy()
orders_copy =orders.copy()
products_copy =products.copy()

# Getting information about tables to be merged

In [4]:
order_items.head()

order_id  product_id  quantity  unit_price
0  1961864118      642612         3      517.03
1  1549769649      378676         1      881.42
2  9185164487      741341         2      923.84
3  9644738826      561860         1      874.78
4  5427684290      602241         2      976.55

In [5]:
feedback.head()

feedback_id    order_id  customer_id  rating  \
0      2234710  1961864118     30065862       4   
1      5450964  1549769649      9573071       3   
2       482108  9185164487     45477575       3   
3      4823104  9644738826     88067569       4   
4      3537464  5427684290     83298567       3   

                          feedback_text feedback_category sentiment  \
0         It was okay, nothing special.          Delivery   Neutral   
1              The order was incorrect.    App Experience  Negative   
2         It was okay, nothing special.    App Experience   Neutral   
3      The product met my expectations.    App Experience   Neutral   
4  Product was damaged during delivery.          Delivery  Negative   

  feedback_date  
0    2024-07-17  
1    2024-05-28  
2    2024-09-23  
3    2023-11-24  
4    2023-11-20

In [28]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2500 entries, 0 to 2499
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        2500 non-null   int64  
 1   customer_name      2500 non-null   object 
 2   email              2500 non-null   object 
 3   phone              2500 non-null   int64  
 4   address            2500 non-null   object 
 5   area               2500 non-null   object 
 6   pincode            2500 non-null   int64  
 7   registration_date  2500 non-null   object 
 8   customer_segment   2500 non-null   object 
 9   total_orders       2500 non-null   int64  
 10  avg_order_value    2500 non-null   float64
dtypes: float64(1), int64(4), object(6)
memory usage: 215.0+ KB


# Created a list of Nigerian areas to change the already existing ones/Replacing the new areas 

In [32]:
##Creating a list of Nigerian Unique_areas
Nigerian_areas=["Lagos",'Port Harcout','Abuja','Uyo','Asaba','Kaduna','Jos','Onitsha','Benin City','Enugu','Calabar','Owerri','Niger','Akure','Warri','Ibadan','Bayelsa','Katsina','Kano','Ogun','Abia','Illorin','Abeokuta'] 

In [79]:
customers["area"].value_counts()
#checking the unique Nigerian areas and their unique count.

area
Port Harcout    121
Benin City      120
Asaba           118
Bayelsa         118
Ogun            116
Abeokuta        115
Uyo             115
Ibadan          115
Kaduna          114
Niger           114
Jos             112
Illorin         112
Abia            112
Lagos           111
Abuja           109
Akure           106
Owerri          104
Warri           100
Onitsha          98
Enugu            96
Calabar          96
Katsina          93
Kano             85
Name: count, dtype: int64

In [34]:
#Replacing the existing areas with the Nigerian areas.

customers['area']=np.random.choice(Nigerian_areas,size=len(customers))

 # MERGING MULTIPLE TABLES TOGETHER/ANSWERING DATA DRIVEN QUESTIONS

## Q1. Are there certain areas that needs better marketing campaigns or delivery improvements?


*  I merged the customer,orders and feedback tables together to be able to analyse the question
*  Changed the areas to realistic areas

In [35]:
customers.merge(orders,"outer","customer_id").merge(feedback,"outer","order_id")

customer_id_x     customer_name                       email  \
0          15808945    Vrinda Khurana         wsekhon@example.com   
1          48281892  Teerth Mukherjee    toordakshesh@example.org   
2          89617089       Kamala Apte     warjasnayar@example.org   
3          44174426     Girindra Nair    wazirsanghvi@example.org   
4          51476157        Abdul Bali    nandinibanik@example.net   
...             ...               ...                         ...   
5323       98016334      Krishna Char  krishnabhavsar@example.net   
5324       98667584        Sara Soman    salvigeetika@example.com   
5325       99347871    Robert Kashyap       gnadkarni@example.net   
5326       99424244    Bhavini Mander   yashasvipalan@example.net   
5327       99643057          Avni Lad    vohratanveer@example.org   

             phone                                    address          area  \
0     918358677006     H.No. 515, Balay Path\nAmravati 503358         Asaba   
1     911236990834  H.No. 706\nLal Path\nVisakhapatnam-582362  Port Harcout   
2     913898334323       H.No. 63\nSarkar Circle\nOrai 521594        Ibadan   
3     915848023986         12/240, Barad Ganj\nHajipur 503025         Lagos   
4     919436539657    62/25\nSarma Circle\nMahbubnagar-707575        Owerri   
...            ...                                        ...           ...   
5323  919448360416           841, Kibe Chowk, Thrissur-092044          Kano   
5324  912270298752    01/760\nNatarajan Path\nParbhani-638480    Benin City   
5325  915345369198       37/115\nKade Nagar\nBaranagar 427420        Ibadan   
5326  916324786919      800\nSubramanian Street\nDhule-015954       Illorin   
5327  915454861960  H.No. 592\nVyas Path, Narasaraopet 538104          Kano   

      pincode registration_date customer_segment  total_orders  ...  \
0      890545        2024-04-10              New             6  ...   
1      691199        2023-04-11              New             6  ...   
2      710095        2024-04-20              New             9  ...   
3      133683        2024-07-02              New            20  ...   
4      928472        2024-08-21         Inactive            10  ...   
...       ...               ...              ...           ...  ...   
5323   326443        2024-08-19          Regular            18  ...   
5324   439446        2024-10-27          Premium             8  ...   
5325   535677        2023-12-01          Regular             3  ...   
5326   184817        2024-03-30          Premium            15  ...   
5327   931414        2024-11-01         Inactive             4  ...   

      payment_method  delivery_partner_id store_id feedback_id customer_id_y  \
0                UPI              58998.0   3943.0   4344466.0    15808945.0   
1             Wallet              97945.0   1987.0    342939.0    48281892.0   
2                UPI              20246.0    974.0   5317004.0    89617089.0   
3             Wallet              48582.0   3184.0   4367429.0    44174426.0   
4             Wallet              44545.0    419.0   2650144.0    51476157.0   
...              ...                  ...      ...         ...           ...   
5323             NaN                  NaN      NaN         NaN           NaN   
5324             NaN                  NaN      NaN         NaN           NaN   
5325             NaN                  NaN      NaN         NaN           NaN   
5326             NaN                  NaN      NaN         NaN           NaN   
5327             NaN                  NaN      NaN         NaN           NaN   

     rating                                      feedback_text  \
0       4.0                   The product met my expectations.   
1       3.0  Nothing to complain about, but nothing excepti...   
2       1.0                  Customer service was not helpful.   
3       5.0                       Very happy with the service!   
4       3.0                        Not worth the price I paid.   
...     ...       

# Assigning the merged table to a name 'Customer_order_feedback'

In [36]:
Customer_order_feedback= customers.merge(orders,"outer","customer_id").merge(feedback,"outer","order_id")

In [12]:
Customer_order_feedback.info()#checking for anomalies in the tables merged

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5328 entries, 0 to 5327
Data columns (total 27 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   customer_id_x           5328 non-null   int64  
 1   customer_name           5328 non-null   object 
 2   email                   5328 non-null   object 
 3   phone                   5328 non-null   int64  
 4   address                 5328 non-null   object 
 5   area                    5328 non-null   object 
 6   pincode                 5328 non-null   int64  
 7   registration_date       5328 non-null   object 
 8   customer_segment        5328 non-null   object 
 9   total_orders            5328 non-null   int64  
 10  avg_order_value         5328 non-null   float64
 11  order_id                5000 non-null   float64
 12  order_date              5000 non-null   object 
 13  promised_delivery_time  5000 non-null   object 
 14  actual_delivery_time    5000 non-null   

# Checking for duplicates

In [13]:
print((Customer_order_feedback["customer_id_x"] ==Customer_order_feedback["customer_id_y"]).all())
##checking if the customer_id_x and customer_id_y is the same


False


# Dropping unnecessary variables/renaming customer_id_y to 'customer_id'

In [21]:
Customer_order_feedback.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5000 entries, 0 to 4999
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   customer_name           5000 non-null   object        
 1   area                    5000 non-null   object        
 2   customer_segment        5000 non-null   object        
 3   total_orders            5000 non-null   int64         
 4   avg_order_value         5000 non-null   float64       
 5   order_id                5000 non-null   float64       
 6   order_date              5000 non-null   object        
 7   promised_delivery_time  5000 non-null   datetime64[ns]
 8   actual_delivery_time    5000 non-null   datetime64[ns]
 9   delivery_status         5000 non-null   object        
 10  order_total             5000 non-null   float64       
 11  payment_method          5000 non-null   object        
 12  delivery_partner_id     5000 non-null   float64      

In [37]:
Customer_order_feedback = Customer_order_feedback.drop(columns =["customer_id_x","pincode", "registration_date","email","phone","address"])
Customer_order_feedback.rename(columns ={"customer_id_y":"customer_id"}, inplace=True)

##drop customer_id_x,pincode,registration dates,email columns and also renamed customer_id_y to "customer_id"

In [15]:
Customer_order_feedback[Customer_order_feedback["order_id"].isna()]
##information concerning the NaNs

customer_name      area customer_segment  total_orders  \
5000           Ekiya Pau     Enugu              New            18   
5001         Jeet Gandhi     Warri          Regular            13   
5002     Chatura Agarwal   Illorin         Inactive             5   
5003  Yadavi Rajagopalan   Calabar         Inactive             4   
5004         Bhavani Rau   Bayelsa          Regular             8   
...                  ...       ...              ...           ...   
5323        Krishna Char  Abeokuta          Regular            18   
5324          Sara Soman   Bayelsa          Premium             8   
5325      Robert Kashyap   Bayelsa          Regular             3   
5326      Bhavini Mander     Abuja          Premium            15   
5327            Avni Lad   Bayelsa         Inactive             4   

      avg_order_value  order_id order_date promised_delivery_time  \
5000           752.78       NaN        NaN                    NaN   
5001          1938.38       NaN        NaN                    NaN   
5002          1323.32       NaN        NaN                    NaN   
5003           999.24       NaN        NaN                    NaN   
5004          1975.21       NaN        NaN                    NaN   
...               ...       ...        ...                    ...   
5323           724.91       NaN        NaN                    NaN   
5324          1072.65       NaN        NaN                    NaN   
5325          1535.03       NaN        NaN                    NaN   
5326           379.80       NaN        NaN                    NaN   
5327          1677.24       NaN        NaN                    NaN   

     actual_delivery_time delivery_status  ...  payment_method  \
5000                  NaN             NaN  ...             NaN   
5001                  NaN             NaN  ...             NaN   
5002                  NaN             NaN  ...             NaN   
5003                  NaN             NaN  ...             NaN   
5004                  NaN             NaN  ...             NaN   
...                   ...             ...  ...             ...   
5323                  NaN             NaN  ...             NaN   
5324                  NaN             NaN  ...             NaN   
5325                  NaN             NaN  ...             NaN   
5326                  NaN             NaN  ...             NaN   
5327                  NaN             NaN  ...             NaN   

     delivery_partner_id  store_id  feedback_id  customer_id  rating  \
5000                 NaN       NaN          NaN          NaN     NaN   
5001                 NaN       NaN          NaN          NaN     NaN   
5002                 NaN       NaN          NaN          NaN     NaN   
5003                 NaN       NaN          NaN          NaN     NaN   
5004                 NaN       NaN          NaN          NaN     NaN   
...                  ...       ...          ...          ...     ...   
5323                 NaN       NaN          NaN          NaN     NaN   
5324                 NaN       NaN          NaN          NaN     NaN   
5325                 NaN       NaN          NaN          NaN     NaN   
5326                 NaN       NaN          NaN          NaN     NaN   
5327                 NaN       NaN          NaN          NaN     NaN   

      feedback_text feedback_category sentiment feedback_date  
5000            NaN               NaN       NaN           NaN  
5001            NaN               NaN       NaN           NaN  
5002            NaN               NaN       NaN           NaN  
5003            NaN               NaN       NaN           NaN  
5004            NaN               NaN       NaN           NaN  
...             ...               ...       ...           ...  
5323            NaN               NaN       NaN           NaN  
5324            NaN               NaN       NaN           NaN  
5325            NaN               NaN       NaN           NaN  
5326            NaN               NaN       NaN        

In [80]:
Customer_order_feedback = Customer_order_feedback.dropna(subset=["order_id"])
##since the 328 has no information,Ill just drop the rows

# Converting promised delivery time and actual delivery time to datetime

In [38]:
Customer_order_feedback["promised_delivery_time"] = pd.to_datetime(Customer_order_feedback["promised_delivery_time"])
Customer_order_feedback["actual_delivery_time"] = pd.to_datetime(Customer_order_feedback["actual_delivery_time"])

# Creating a variable 'delivery status'/Checking the counts of delivery status Per Area

In [41]:
pivot_delivery = delivery_status_counts.pivot(index="area", columns="delivery_status", values="count").fillna(0)
pivot_delivery

delivery_status  On Time  Significantly Delayed  Slightly Delayed
area                                                             
Abeokuta             153                     31                52
Abia                 165                     23                43
Abuja                153                     23                41
Akure                152                     22                51
Asaba                164                     24                57
Bayelsa              142                     19                41
Benin City           158                     24                47
Calabar              145                     11                37
Enugu                131                     22                43
Ibadan               177                     25                41
Illorin              135                     23                45
Jos                  186                     17                42
Kaduna               182                     25                40
Kano                 113                     13                27
Katsina              113                     16                51
Lagos                147                     24                54
Niger                161                     26                47
Ogun                 149                     30                55
Onitsha              149                     13                41
Owerri               129                     19                50
Port Harcout         151                     25                47
Uyo                  187                     19                45
Warri                128                     19                40

In [40]:
delivery_status_counts = Customer_order_feedback.groupby(
    ["area", "delivery_status"]
).size().reset_index(name="count")

In [42]:
## Grouping by area and delivery_status to get counts of delivery per area
delivery_status_counts = Customer_order_feedback.groupby(
    ["area", "delivery_status"]
).size().reset_index(name="count")


delivery_status_pivot = delivery_status_counts.pivot(
    index="area", columns="delivery_status", values="count"
).fillna(0)

# Adding a total column
delivery_status_pivot["total"] = delivery_status_pivot.sum(axis=1)

# Calculate percentage columns
for col in ["On Time", "Slightly Delayed", "Significantly Delayed"]:
    delivery_status_pivot[col + "_%"] = (
        delivery_status_pivot[col] / delivery_status_pivot["total"] * 100
    ).round(0).astype(str) + "%"  # format as percentage string

# Reset index to make area a column again
Delivery_improvement = delivery_status_pivot.reset_index()
Delivery_improvement



delivery_status          area  On Time  Significantly Delayed  \
0                    Abeokuta      153                     31   
1                        Abia      165                     23   
2                       Abuja      153                     23   
3                       Akure      152                     22   
4                       Asaba      164                     24   
5                     Bayelsa      142                     19   
6                  Benin City      158                     24   
7                     Calabar      145                     11   
8                       Enugu      131                     22   
9                      Ibadan      177                     25   
10                    Illorin      135                     23   
11                        Jos      186                     17   
12                     Kaduna      182                     25   
13                       Kano      113                     13   
14                    Katsina      113                     16   
15                      Lagos      147                     24   
16                      Niger      161                     26   
17                       Ogun      149                     30   
18                    Onitsha      149                     13   
19                     Owerri      129                     19   
20               Port Harcout      151                     25   
21                        Uyo      187                     19   
22                      Warri      128                     19   

delivery_status  Slightly Delayed  total On Time_% Slightly Delayed_%  \
0                              52    236     65.0%              22.0%   
1                              43    231     71.0%              19.0%   
2                              41    217     71.0%              19.0%   
3                              51    225     68.0%              23.0%   
4                              57    245     67.0%              23.0%   
5                              41    202     70.0%              20.0%   
6                              47    229     69.0%              21.0%   
7                              37    193     75.0%              19.0%   
8                              43    196     67.0%              22.0%   
9                              41    243     73.0%              17.0%   
10                             45    203     67.0%              22.0%   
11                             42    245     76.0%              17.0%   
12                             40    247     74.0%              16.0%   
13                             27    153     74.0%              18.0%   
14                             51    180     63.0%              28.0%   
15                             54    225     65.0%              24.0%   
16                             47    234     69.0%              20.0%   
17                             55    234     64.0%              24.0%   
18                             41    203     73.0%              20.0%   
19                             50    198     65.0%              25.0%   
20                             47    223     68.0%              21.0%   
21                             45    251     75.0%              18.0%   
22                             40    187     68.0%              21.0%   

delivery_status Significantly Delayed_%  
0                                 13.0%  
1                                 10.0%  
2                                 11.0%  
3                                 10.0%  
4                                 10.0%  
5                                  9.0%  
6                                 10.0%  
7                                  6.0%  
8                                 11.0%  
9                                 10.0%  
10                                11.0%  
11                                 7.0%  
12                                10.0%  
13                                 8.0%  
14                                 9.0%  
15            

In [ ]:
Delivery_improvement.to_csv(r"C:\Users\ROSZ\Desktop\Data Analysisr\BLINKIT/Delivery_improvement.csv",index=False)

#  Observations/Analysis on Q1

* After merging the three tables together using the outer join,I observed that there are two customer_id variables which are categorised into customer_id_x and customer_id_y and the row number in both of them do not match.Why the mismatch?customer_id_x → comes from customers table (all registered customers). That’s 5328 rows.customer_id_y → comes from orders/feedback table (only customers who actually placed an order and gave feedback). That’s 5000 rows.
This means the remaining 328 customers exist  but never placed an order or gave feedback hence the mismatch
* I dropped customer_id_x and renamed the customer_id_y to customer_id
* I dropped variables that may not be necessary for the first analysis,like registration dates,pincodes,address etc.
* * To answer the question concerning delivery,from the dtypes,pandas automatically changed delivery date to object,to be able to explore the data there,I converted it to datetime.
* From the table above,it shows the percentage of deliveries (on-time,slightly delayed and significantly delayed) in the unique areas,this shows the areas that needs delivery improvements.

## Q2. WHICH CUSTOMER SEGMENT BRINGS IN THE MOST REVENUE OR ORDERS FREQUENTLY?

In [43]:
customer_segment_analysis = Customer_order_feedback.groupby("customer_segment").agg({
    "order_id": "count",      # number of orders
    "order_total": "sum"      # total revenue
})

(customer_segment_analysis.sort_values(by="order_total", ascending=False)).reset_index()


customer_segment  order_id  order_total
0          Regular      1320   2890148.66
1              New      1222   2795854.48
2          Premium      1268   2731329.71
3         Inactive      1190   2591975.65

In [ ]:
customer_segment_analysis.to_csv(r"C:\Users\ROSZ\Desktop\Data Analysisr\BLINKIT/customer_segment_analysis.csv")
##exporting file to csv

## Analysis From Q2.
* From the table above,it shows that the REGULAR customer segment orders more frequently and they have  highest revenue as well.

## Q3. WHAT PRODUCT GENERATES THE HIGHEST REVENUE
* To analyse this question, I'll merge products table and order_items tables
* Create a revenue variable by multiplying quantity by unit_price

In [ ]:
order_items.merge(products,'outer','product_id')

# Assigning merged table to a dataframe 'Products_order_items'

In [59]:
Products_order_items=order_items.merge(products,'outer','product_id')

In [60]:
Products_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   order_id           5000 non-null   int64  
 1   product_id         5000 non-null   int64  
 2   quantity           5000 non-null   int64  
 3   unit_price         5000 non-null   float64
 4   product_name       5000 non-null   object 
 5   category           5000 non-null   object 
 6   brand              5000 non-null   object 
 7   price              5000 non-null   float64
 8   mrp                5000 non-null   float64
 9   margin_percentage  5000 non-null   int64  
 10  shelf_life_days    5000 non-null   int64  
 11  min_stock_level    5000 non-null   object 
 12  max_stock_level    5000 non-null   int64  
dtypes: float64(3), int64(6), object(4)
memory usage: 507.9+ KB


# Checking for duplicates/dropping duplicates

In [61]:
(Products_order_items["unit_price"] == Products_order_items["price"]).all()

#checking for duplicates in columns unit_price and price vaiables.

np.True_

In [62]:
Products_order_items.drop(columns =('price'))

##dropping the duplicate column

order_id  product_id  quantity  unit_price               product_name  \
0     4976389251        4452         1      169.88  Matching Bra & Panty sets   
1      511736499        4452         1      169.88  Matching Bra & Panty sets   
2      877883470        4452         2      169.88  Matching Bra & Panty sets   
3      637944734        4452         3      169.88  Matching Bra & Panty sets   
4     9581592141        4452         3      169.88  Matching Bra & Panty sets   
...          ...         ...       ...         ...                        ...   
4995   948345459      993331         1      102.44                   Bralette   
4996   913881994      993331         2      102.44                   Bralette   
4997  6421432903      993331         2      102.44                   Bralette   
4998   996939836      993331         2      102.44                   Bralette   
4999  9251851237      993331         3      102.44                   Bralette   

           category                  brand     mrp  margin_percentage  \
0     Lingerie Sets           Morar-Mistry  242.69                 30   
1     Lingerie Sets           Morar-Mistry  242.69                 30   
2     Lingerie Sets           Morar-Mistry  242.69                 30   
3     Lingerie Sets           Morar-Mistry  242.69                 30   
4     Lingerie Sets           Morar-Mistry  242.69                 30   
...             ...                    ...     ...                ...   
4995  Lingerie Sets  Parmer, Gola and Yogi  120.52                 15   
4996  Lingerie Sets  Parmer, Gola and Yogi  120.52                 15   
4997  Lingerie Sets  Parmer, Gola and Yogi  120.52                 15   
4998  Lingerie Sets  Parmer, Gola and Yogi  120.52                 15   
4999  Lingerie Sets  Parmer, Gola and Yogi  120.52                 15   

      shelf_life_days min_stock_level  max_stock_level  
0                 365              27              100  
1                 365              27              100  
2                 365              27              100  
3                 365              27              100  
4                 365              27              100  
...               ...             ...              ...  
4995              365              28               86  
4996              365              28               86  
4997              365              28               86  
4998              365              28               86  
4999              365              28               86  

[5000 rows x 12 columns]

In [63]:
Products_order_items['Revenue']=Products_order_items['quantity']*Products_order_items['unit_price']
(Products_order_items['Revenue'])

       
##multiplying quantity and unit_price,all from the merged table to get revenue

0       169.88
1       169.88
2       339.76
3       509.64
4       509.64
         ...  
4995    102.44
4996    204.88
4997    204.88
4998    204.88
4999    307.32
Name: Revenue, Length: 5000, dtype: float64

In [130]:
Products_order_items = Products_order_items.groupby("product_name").agg({
    "quantity": "sum",   # number of orders
    "unit_price": "mean",       # average price
    "Revenue": "sum"      # total revenue
})

(Products_order_items.sort_values(by="Revenue", ascending=False)).reset_index()

product_name  quantity  unit_price    Revenue
0             Bikini Panties      1256  600.798155  751833.88
1  Matching Bra & Panty sets      1314  468.284202  611919.89
2                   Bralette      1010  518.461780  532945.35
3                    Pajamas       844  616.490357  528466.88
4                 Body Suits      1092  424.698230  465658.07
5         High-waist Panties      1012  437.037422  447637.58
6                  Lace Sets       921  469.787223  435365.27
7                     Robes        898  482.825919  432453.98
8               Body Shapers       827  490.401411  407543.72
9                 Night gown       860  417.971442  358590.81

In [144]:
Products_order_items.to_csv(r"C:\Users\ROSZ\Desktop\Data Analysisr\BLINKIT/Products_order_items.csv")

## Observations/Analysis For Q3
* After merging products and order_items tables together,I noticed there are duplicates, unit_price and price are the same figures,I dropped the price column. 
* The analysis shows the products that generates the highest revenue according to their order and "Bikini Panties" topping the highest product.

## Q4.WHAT TIME OF DAY DO CUSTOMERS MAKE THE MOST ORDER?

* To analyse this,I'll convert order_date in order table to datetime datatype from the already existed merged table 'Customer_order_feedback'
* Convert order_date dtype to datetime
* Extract hour from order_time
* Create time_of_day variable

In [65]:
Customer_order_feedback["order_date"] = pd.to_datetime(Customer_order_feedback["order_date"])

# Collecting information from already existing merged table

In [66]:
Customer_order_feedback.info() ##checking info about table

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5328 entries, 0 to 5327
Data columns (total 23 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   customer_name           5328 non-null   object        
 1   area                    5328 non-null   object        
 2   customer_segment        5328 non-null   object        
 3   total_orders            5328 non-null   int64         
 4   avg_order_value         5328 non-null   float64       
 5   order_id                5000 non-null   float64       
 6   order_date              5000 non-null   datetime64[ns]
 7   promised_delivery_time  5000 non-null   datetime64[ns]
 8   actual_delivery_time    5000 non-null   datetime64[ns]
 9   delivery_status         5000 non-null   object        
 10  order_total             5000 non-null   float64       
 11  payment_method          5000 non-null   object        
 12  delivery_partner_id     5000 non-null   float64 

# Extracting order hour from order date variable

In [67]:
Customer_order_feedback["order_hour"] = Customer_order_feedback["order_date"].dt.hour
##extracting the hours from order_date

# Creating time of day variable

In [145]:
def get_time_of_day(hour):
    if  0 <= hour < 4:
        return "Dawn"
    elif 7 < hour < 12:
        return "Morning"
    elif 12 < hour < 16:
        return "Afternoon"
    elif 16 < hour < 20:
        return "Evening"
    else:
        return "Night"

Customer_order_feedback["time_of_day"] = Customer_order_feedback["order_hour"].apply(get_time_of_day)
##Creating time of day variable

In [146]:
subset = Customer_order_feedback[["time_of_day", "order_date"]]
(subset)


time_of_day          order_date
0          Night 2024-10-23 05:23:29
1           Dawn 2023-04-02 03:45:11
2           Dawn 2024-05-23 03:21:47
3          Night 2023-06-09 12:10:20
4          Night 2023-04-30 20:52:21
...          ...                 ...
4995       Night 2023-05-23 04:12:41
4996   Afternoon 2023-04-18 13:13:06
4997        Dawn 2024-04-21 00:03:15
4998   Afternoon 2023-12-27 14:02:18
4999     Evening 2023-09-22 17:07:28

[5000 rows x 2 columns]

# Checking the time of day orders was most made

In [147]:
Time_analysis = Customer_order_feedback.groupby(["time_of_day"])["order_id"].count().reset_index()
Time_analysis.rename(columns={'order_id':'order_count'},inplace=True)  #renaming variable for clarity

Time_analysis= Time_analysis.sort_values(by='order_count',ascending=False).reset_index()
(Time_analysis)

index time_of_day  order_count
0      4       Night         2075
1      1        Dawn          819
2      3     Morning          808
3      2     Evening          658
4      0   Afternoon          640

## Analysis for Q4
* As seen in the table above,customers make their most orders at Night

In [148]:
Time_analysis.to_csv(r"C:\Users\ROSZ\Desktop\Data Analysisr\BLINKIT/Time_analysis.csv",index=False)

# Q5. ANALYZE THE MARKETING CHANNEL WITH HIGHEST CONVERSION

* This will be analysed using the conversions variable from the marketing table

In [71]:
marketing.head()

campaign_id      campaign_name        date target_audience channel  \
0       548299  New User Discount  2024-11-05         Premium     App   
1       390914    Weekend Special  2024-11-05        Inactive     App   
2       834385     Festival Offer  2024-11-05        Inactive   Email   
3       241523         Flash Sale  2024-11-05        Inactive     SMS   
4       595111   Membership Drive  2024-11-05       New Users   Email   

   impressions  clicks  conversions    spend  revenue_generated  roas  
0         3130     163           78  1431.85            4777.75  3.60  
1         3925     494           45  4506.34            6238.11  2.98  
2         7012     370           78  4524.23            2621.00  2.95  
3         1115     579           86  3622.79            2955.00  2.84  
4         7172     795           54  2888.99            8951.81  2.22

In [72]:
Channel_audience_conversion = marketing.groupby(
    ['channel', 'target_audience']
).agg({
    'conversions': 'sum' #total number of conversions
}).reset_index()

most_responsive_channel = (Channel_audience_conversion.sort_values(
    ['channel', 'conversions'], ascending=[True, False]
).groupby('channel').head())

most_responsive_channel

channel target_audience  conversions
0            App             All        20371
1            App        Inactive        19364
2            App       New Users        19303
3            App         Premium        16154
6          Email       New Users        18872
4          Email             All        18789
7          Email         Premium        18712
5          Email        Inactive        18298
11           SMS         Premium        19256
9            SMS        Inactive        18878
10           SMS       New Users        18613
8            SMS             All        16488
15  Social Media         Premium        19481
14  Social Media       New Users        18968
12  Social Media             All        18889
13  Social Media        Inactive        17602

# Analysis on Q5

* From the table above, I grouped each channel and their targeted audience and I deduced that the "APP" had the most conversion.

In [ ]:
most_responsive_channel.to_csv(r"C:\Users\ROSZ\Desktop\Data Analysisr\BLINKIT/most_responsive_channel.csv",index=False)

# Q6. THE LEVEL OF SATISFATION FROM CUSTOMERS

In [73]:
Customer_order_feedback.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5328 entries, 0 to 5327
Data columns (total 23 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   customer_name           5328 non-null   object        
 1   area                    5328 non-null   object        
 2   customer_segment        5328 non-null   object        
 3   total_orders            5328 non-null   int64         
 4   avg_order_value         5328 non-null   float64       
 5   order_id                5000 non-null   float64       
 6   order_date              5000 non-null   datetime64[ns]
 7   promised_delivery_time  5000 non-null   datetime64[ns]
 8   actual_delivery_time    5000 non-null   datetime64[ns]
 9   delivery_status         5000 non-null   object        
 10  order_total             5000 non-null   float64       
 11  payment_method          5000 non-null   object        
 12  delivery_partner_id     5000 non-null   float64 

In [74]:
rating_dist = Customer_order_feedback["rating"].value_counts(normalize=True) * 100
(rating_dist)
#percentage of rating for all customers.

rating
4.0    34.16
3.0    27.96
5.0    16.32
1.0    10.80
2.0    10.76
Name: proportion, dtype: float64

In [75]:
Customer_satisfaction =Customer_order_feedback .groupby("customer_segment").agg({
    'rating': 'mean' #avg rating
        }).reset_index()
Customer_satisfaction

customer_segment    rating
0         Inactive  3.366387
1              New  3.339607
2          Premium  3.320978
3          Regular  3.351515

# Converting 'rating' from float to interger

In [ ]:
Customer_order_feedback["rating"] = Customer_order_feedback["rating"].astype(int)

# Creating 'Satisfaction table' to be able to analyse its probability

In [77]:
# creatinng satisfaction table
def get_satisfaction_table(rating):
    if rating == 1:
        return "Very Poor"
    elif rating == 2:
        return "Poor"
    elif rating == 3:
        return "Neutral"
    elif rating == 4:
        return "Satisfied"
    elif rating == 5:
        return "Very Satisfied"
    else:
        return "Unknown"

Customer_order_feedback["satisfaction_table"] = Customer_order_feedback["rating"].apply(get_satisfaction_table)

In [82]:
subset2 = Customer_order_feedback[["satisfaction_table", "rating"]]
subset2

satisfaction_table  rating
0             Satisfied     4.0
1               Neutral     3.0
2             Very Poor     1.0
3        Very Satisfied     5.0
4               Neutral     3.0
...                 ...     ...
4995          Satisfied     4.0
4996            Neutral     3.0
4997     Very Satisfied     5.0
4998     Very Satisfied     5.0
4999            Neutral     3.0

[5000 rows x 2 columns]

## Getting the probablity of the customer segment's level of performance

## Converting ratings 4 and 5 to 'is_satisfied'

In [153]:
Customer_order_feedback['is_satisfied']=Customer_order_feedback['rating'].isin([4,5])


In [162]:
segment_satisfaction_rate = (
    Customer_order_feedback.groupby("customer_segment")["is_satisfied"]
    .mean() * 100
).reset_index()

segment_satisfaction_rate


customer_segment  is_satisfied
0         Inactive     50.924370
1              New     50.000000
2          Premium     49.763407
3          Regular     51.212121

In [ ]:
segment_satisfaction_rate.to_csv(r"C:\Users\ROSZ\Desktop\Data Analysisr\BLINKIT/segment_satisfaction_rate.csv",index=False)

In [ ]:
rating_dist.to_csv(r"C:\Users\ROSZ\Desktop\Data Analysisr\BLINKIT/rating_percent_distri.csv")

## Explanation on Q6

* The rating distribution table shows the percentage of rating on all customers which is 34%, it means that 34% customers gave an average experience that is 4 stars.However,when broken down by customer segment,all groups average around 3.32 to 3.37 ,showing moderate satisfaction.
* The table also shows 


In [85]:
# 1. Create a pivot table for counts of each rating by customer segment
rating_counts = pd.crosstab(
    Customer_order_feedback["customer_segment"],
    Customer_order_feedback["rating"]
)

# 2. Rename the columns so they look like rating_1, rating_2, etc.
rating_counts = rating_counts.rename(
    columns={i: f"rating_{int(i)}" for i in rating_counts.columns}
)

# 3. Add average rating for each segment
avg_ratings = (
    Customer_order_feedback.groupby("customer_segment")["rating"]
    .mean()
    .rename("avg_rating")
)

# 4. Combine into one table
segment_summary = rating_counts.join(avg_ratings)

segment_summary


rating_1  rating_2  rating_3  rating_4  rating_5  avg_rating
customer_segment                                                              
Inactive               133       113       338       397       209    3.366387
New                    132       127       352       416       195    3.339607
Premium                138       155       344       424       207    3.320978
Regular                137       143       364       471       205    3.351515

## More Explantion of Q6.
* From the table above,regular customers have the highest number when summed together i.e the ratings 4 and 5 making them the most consistently satisfied segment.
* Premium customers hover around average satisfaction but that does not equal higher satisfaction.
* The other two customer segments are the least satisfied.

## Q7.  WHICH AREAS HAVE THE HIGHEST NUMBER OF ORDERS?

In [133]:
Highest_orders_area=Customer_order_feedback.groupby('area').agg({
    'total_orders': 'sum' #total sum of orders
})
Highest_orders_area.sort_values(by='total_orders',ascending=False).reset_index()

area  total_orders
0            Uyo          2752
1            Jos          2642
2          Asaba          2633
3         Kaduna          2632
4         Ibadan          2619
5          Niger          2561
6     Benin City          2525
7       Abeokuta          2390
8          Lagos          2381
9           Ogun          2331
10          Abia          2327
11  Port Harcout          2290
12         Abuja          2265
13       Onitsha          2261
14       Illorin          2196
15         Akure          2160
16       Calabar          2098
17         Enugu          2075
18        Owerri          2048
19       Bayelsa          1974
20         Warri          1872
21       Katsina          1856
22          Kano          1523

## Explantion from Q7.
* From the table,it shows areas that tops the highest in their order, with UYO being the top of the list.


In [136]:
Highest_orders_area.to_csv(r"C:\Users\ROSZ\Desktop\Data Analysisr\BLINKIT/Highest_orders_area.csv")
# exporting table to csv file

# Q8. WHAT IS THE MOST PREFERRED PAYMENT METHOD?

# Getting the most preferred mode of payment in percentage for the general customers

In [87]:
Payment_pref_percent = Customer_order_feedback["payment_method"].value_counts(normalize=True) * 100
Payment_pref_percent
## Getting the percentage of most preferred payment method.

payment_method
Card      25.70
Cash      25.14
Wallet    24.88
UPI       24.28
Name: proportion, dtype: float64

## Converting series to dataframe

In [88]:
Payment_pref_percentage=pd.DataFrame(Payment_pref_percent).reset_index()

In [89]:
Payment_pref_percentage

payment_method  proportion
0           Card       25.70
1           Cash       25.14
2         Wallet       24.88
3            UPI       24.28

In [90]:
Payment_pref_percentage.to_csv(r"C:\Users\ROSZ\Desktop\Data Analysisr\BLINKIT/Payment_pref_percentage.csv")
#exporting to csv

# Getting the most preferred mode of payment by areas

In [91]:
Payment_pref_by_area = Customer_order_feedback.pivot_table(
    index="area",
    columns="payment_method",
    values="order_id",
    aggfunc="count",
    fill_value=0
)
Payment_pref_by_area

payment_method  Card  Cash  UPI  Wallet
area                                   
Abeokuta          59    68   51      58
Abia              53    67   57      54
Abuja             61    50   60      46
Akure             64    60   45      56
Asaba             61    76   53      55
Bayelsa           45    59   48      50
Benin City        57    55   61      56
Calabar           57    46   39      51
Enugu             54    43   50      49
Ibadan            63    58   54      68
Illorin           47    46   52      58
Jos               71    68   50      56
Kaduna            59    70   62      56
Kano              31    48   37      37
Katsina           43    41   48      48
Lagos             70    51   49      55
Niger             45    57   74      58
Ogun              58    52   53      71
Onitsha           49    49   55      50
Owerri            52    46   48      52
Port Harcout      63    48   53      59
Uyo               64    59   68      60
Warri             59    40   47      41

In [92]:
Payment_pref_by_area.to_csv(r"C:\Users\ROSZ\Desktop\Data Analysisr\BLINKIT/Payment_pref_by_area.csv")
#exporting file to csv

# Analysis on Q8.
* From the table above,it shows that customers most preferred payment method is card giving about 26% but the other mode of payment are almost at the same range.
* The table also shows the count of node of payment that the unique areas use.

## Q9. HOW IS THE BUSINESS PERFORMANCE OVERTIME

To analyse this question, I'll merge orders and order_items tables together.

# Merging orders and order items table together

In [105]:
orders.merge(order_items,'outer','order_id')

order_id  customer_id           order_date promised_delivery_time  \
0          60465     15808945  2024-10-23 05:23:29    2024-10-23 05:34:29   
1        2237858     48281892  2023-04-02 03:45:11    2023-04-02 04:00:11   
2        3101265     89617089  2024-05-23 03:21:47    2024-05-23 03:37:47   
3        5120698     44174426  2023-06-09 12:10:20    2023-06-09 12:25:20   
4        5512907     51476157  2023-04-30 20:52:21    2023-04-30 21:11:21   
...          ...          ...                  ...                    ...   
4995  9993542681     42589311  2023-05-23 04:12:41    2023-05-23 04:26:41   
4996  9996182478     39976965  2023-04-18 13:13:06    2023-04-18 13:30:06   
4997  9997121955     46623192  2024-04-21 00:03:15    2024-04-21 00:19:15   
4998  9997994965     13796199  2023-12-27 14:02:18    2023-12-27 14:20:18   
4999  9998297778      7728406  2023-09-22 17:07:28    2023-09-22 17:26:28   

     actual_delivery_time   delivery_status  order_total payment_method  \
0     2024-10-23 05:39:29           On Time       589.47            UPI   
1     2023-04-02 04:03:11           On Time      3835.26         Wallet   
2     2024-05-23 03:40:47           On Time      3567.73            UPI   
3     2023-06-09 12:40:20  Slightly Delayed      3064.36         Wallet   
4     2023-04-30 21:26:21  Slightly Delayed       861.53         Wallet   
...                   ...               ...          ...            ...   
4995  2023-05-23 04:41:41  Slightly Delayed      3564.26            UPI   
4996  2023-04-18 13:30:06           On Time       340.36            UPI   
4997  2024-04-21 00:16:15           On Time      1756.40            UPI   
4998  2023-12-27 14:17:18           On Time      2780.28           Card   
4999  2023-09-22 17:27:28           On Time      2833.20           Card   

      delivery_partner_id  store_id  product_id  quantity  unit_price  
0                   58998      3943      361361         2      826.21  
1                   97945      1987      585299         1      799.46  
2                   20246       974      300159         1       44.35  
3                   48582      3184      187399         1       53.49  
4                   44545       419      659290         3      319.34  
...                   ...       ...         ...       ...         ...  
4995                73532      4596      880510         1      951.43  
4996                48495      2861      592784         3      280.35  
4997                27103      6589      739534         2      994.56  
4998                28092      4117      461804         2      310.69  
4999                15629      5907      709916         3       33.76  

[5000 rows x 13 columns]

# Assigning the merged table to a dataframe 'Business_perf'

In [106]:
Business_perf=orders.merge(order_items,'outer','order_id')

In [107]:
Business_perf.info() #checking info about the table to be sure the data is cleaned

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   order_id                5000 non-null   int64  
 1   customer_id             5000 non-null   int64  
 2   order_date              5000 non-null   object 
 3   promised_delivery_time  5000 non-null   object 
 4   actual_delivery_time    5000 non-null   object 
 5   delivery_status         5000 non-null   object 
 6   order_total             5000 non-null   float64
 7   payment_method          5000 non-null   object 
 8   delivery_partner_id     5000 non-null   int64  
 9   store_id                5000 non-null   int64  
 10  product_id              5000 non-null   int64  
 11  quantity                5000 non-null   int64  
 12  unit_price              5000 non-null   float64
dtypes: float64(2), int64(6), object(5)
memory usage: 507.9+ KB


# Extracting order_year and month from order_date variable

In [108]:
Business_perf['order_date']=pd.to_datetime(orders['order_date'])
##converting order_date to datetime dtype

In [109]:
Business_perf['order_year'] =Business_perf['order_date'].dt.year
#extracting year from order_date

In [110]:
Business_perf['order_month'] =Business_perf['order_date'].dt.strftime("%b")  
# extracting month name

# Creating a variable 'Revenue' from unit_price and qauntity

In [111]:
Business_perf['Revenue']=Business_perf['unit_price'] * Business_perf['quantity']
Business_perf['Revenue']

#Multiplying unit_price by quantity to get the revenue.

0       1652.42
1        799.46
2         44.35
3         53.49
4        958.02
         ...   
4995     951.43
4996     841.05
4997    1989.12
4998     621.38
4999     101.28
Name: Revenue, Length: 5000, dtype: float64

# Grouping the order_year and order_month by the level of Revenue in year and month

In [112]:
Yearly_perf=Business_perf.groupby('order_year').agg({
    'Revenue': 'sum', #total revenue
    'order_id': 'count' #total number of orders
     })                 
Yearly_perf.sort_values(by='order_year', ascending=False).reset_index()

order_year     Revenue  order_id
0        2024  2511850.78      2556
1        2023  2460564.65      2444

In [113]:
Yearly_perf.to_csv(r"C:\Users\ROSZ\Desktop\Data Analysisr\BLINKIT/Yearly_bizperf.csv")

In [114]:
Monthly_perf=Business_perf.groupby(['order_year','order_month']).agg({
        'Revenue': 'sum', #total revenue
        'order_id': 'count' #total number of orders
     })
Monthly_perf.sort_values(by='order_year', ascending=False).reset_index()

order_year order_month    Revenue  order_id
0         2024         Mar  265941.90       251
1         2024         May  272080.69       263
2         2024         Apr  243637.01       241
3         2024         Feb  238378.09       252
4         2024         Aug  248588.70       251
5         2024         Oct  230150.33       247
6         2024         Nov   39598.59        30
7         2024         Sep  229385.93       247
8         2024         Jan  252252.00       270
9         2024         Jul  242310.05       256
10        2024         Jun  249527.49       248
11        2023         Apr  264768.33       238
12        2023         Aug  294308.20       285
13        2023         Jun  227865.86       232
14        2023         Jul  238538.98       244
15        2023         Dec  274849.60       268
16        2023         Sep  244489.74       262
17        2023         Mar  118818.39       120
18        2023         May  289647.31       276
19        2023         Nov  273567.73       265
20        2023         Oct  233710.51       254

In [127]:
Monthly_perf.to_csv(r"C:\Users\ROSZ\Desktop\Data Analysisr\BLINKIT/Monthly_bizperf.csv")

Monthly_perf = Monthly_perf.reset_index()
Monthly_perf

## Creating a variable 'year_month'
# This is done so as to get the quarter of a particular year.

In [123]:
Monthly_perf["year_month"] = pd.to_datetime(
    Monthly_perf["order_year"].astype(str) + "-" + Monthly_perf["order_month"].astype(str) + "-01"
)
Monthly_perf

index  order_year order_month    Revenue  order_id year_month
0       0        2023         Apr  264768.33       238 2023-04-01
1       1        2023         Aug  294308.20       285 2023-08-01
2       2        2023         Dec  274849.60       268 2023-12-01
3       3        2023         Jul  238538.98       244 2023-07-01
4       4        2023         Jun  227865.86       232 2023-06-01
5       5        2023         Mar  118818.39       120 2023-03-01
6       6        2023         May  289647.31       276 2023-05-01
7       7        2023         Nov  273567.73       265 2023-11-01
8       8        2023         Oct  233710.51       254 2023-10-01
9       9        2023         Sep  244489.74       262 2023-09-01
10     10        2024         Apr  243637.01       241 2024-04-01
11     11        2024         Aug  248588.70       251 2024-08-01
12     12        2024         Feb  238378.09       252 2024-02-01
13     13        2024         Jan  252252.00       270 2024-01-01
14     14        2024         Jul  242310.05       256 2024-07-01
15     15        2024         Jun  249527.49       248 2024-06-01
16     16        2024         Mar  265941.90       251 2024-03-01
17     17        2024         May  272080.69       263 2024-05-01
18     18        2024         Nov   39598.59        30 2024-11-01
19     19        2024         Oct  230150.33       247 2024-10-01
20     20        2024         Sep  229385.93       247 2024-09-01

In [124]:
Monthly_perf = Monthly_perf.sort_values("year_month")

## Analysis on Q9

* From the table above,it shows that the business perfomance across the years i.e 2023 and 2024 of which 2023 topped the highest with a total revenue of 2511850.78 slightly different from that of 2023.It only shows that there has been an improvement in the business overtime.
Consequently,the table also shows the business performance across the different months of the 2023 and 2024 making it easier to see the month at which revenue was made the most.